orchagent/portsorch.cpp

```diff
@@ -1177,6 +1180,69 @@ bool PortsOrch::initPort(const string &alias, const set<int> &lane_set)
     return true;
 }
 
+bool PortsOrch::bake()
+{
+    SWSS_LOG_ENTER();
+
+    // Check the APP_DB port table for warm reboot
+    vector<FieldValueTuple> tuples;
+    bool foundPortConfigDone = m_portTable->get("PortConfigDone", tuples);
+    SWSS_LOG_NOTICE("foundPortConfigDone = %d", foundPortConfigDone);
+
+    bool foundPortInitDone = m_portTable->get("PortInitDone", tuples);
+    SWSS_LOG_NOTICE("foundPortInitDone = %d", foundPortInitDone);
+
+    vector<string> keys;
+    m_portTable->getKeys(keys);
+    SWSS_LOG_NOTICE("m_portTable->getKeys %zd", keys.size());
+
+    if (!foundPortConfigDone || !foundPortInitDone)
+    {
+        SWSS_LOG_NOTICE("No port table, fallback to cold start");
+        cleanPortTable(keys);
+        return false;
+    }
+
+    doPortConfigDoneTask(tuples);
+    if (m_portCount != keys.size() - 2)
```

Comment:
After "PortConfigDone",   for some platform, port may be removed with  removePort(it->second),  could it cause this check invalid in case of warm start?

     if (platform && (strstr(platform, BFN_PLATFORM_SUBSTRING) || strstr(platform, MLNX_PLATFORM_SUBSTRING)))
                        {
                            if (!removePort(it->second))
                            {
                                throw runtime_error("PortsOrch initialization failure.");
                            }
                        }

Reply:
removePort() will never del entry in PORT_TABLE. So the answer is no.

---
In reply to: [207439672](https://github.com/Azure/sonic-swss/pull/551#discussion_r207439672) [](ancestors = 207439672)

Reply:
In your change, doPortConfigDoneTask() is also called at construct phase,   
having m_portConfigDone set prematurely will change the processing flow of void PortsOrch::doPortTask(Consumer &consumer).

```
            /* Once all ports received, go through the each port and perform appropriate actions:
             * 1. Remove ports which don't exist anymore
             * 2. Create new ports
             * 3. Initialize all ports
             */
            if (m_portConfigDone && (m_lanesAliasSpeedMap.size() == m_portCount))     <--- will be false before all of the ports handled.
            {
            }
            
            if (!m_portConfigDone)              <------- m_portConfigDone is true. it continue the flow.
            {
                it = consumer.m_toSync.erase(it);      <----- While if m_portConfigDone is false, the request will be deleted, not good for the one phase warm restore.
                continue;
            }
            
            ........
            Port p;
            if (!getPort(alias, p))
            {
                SWSS_LOG_ERROR("Failed to get port id by alias:%s", alias.c_str());   <--- will hit here, since initPort not called yet.
            }
```
            
Why do you want to mess with the port init orders?  We might hit more subtle sequence problems.             

This also reminds me that the previous switching from ProducerState/ConsumerState to Producer/Consumer for portTable might have some potential issue for cold boot,
 the processing after "PortConfigDone" will not get all fields of the port and make the config like fec, AN void.  The old  ProducerState/ConsumerState happened to get all
 fields of a port key every time.

Reply:
You are correct. I removed calling doPortConfigDoneTask in ctor

---
In reply to: [207479060](https://github.com/Azure/sonic-swss/pull/551#discussion_r207479060) [](ancestors = 207479060)